File: Week1/Bai1/Bai1.cpp
#include <bits/stdc++.h>

using namespace std;

// O(n)
int divide1( int a, int b){
	
	// dau cua phep chia
	int sign = ((a< 0) ^ (b < 0)) ? -1 : 1;
	
	a = abs(a);
	b = abs(b);
	
	

	if( b == 0){
		cout << "Ban nhap sai, b phai khac 0" << endl;
		return 0;
	}


	int quotient = 0;	// thuong
	
	while( a >= b){
		
		a -= b;
		quotient = quotient + 1;
	}
	
	cout << sign*quotient << endl;
}

// O(logn)
voiid divide2(int a, int b){

	// dau cua phep chia
	int sign = ((a< 0) ^ (b < 0)) ? -1 : 1;
	
	a = abs(a);
	b = abs(b);
	

	if( b == 0){
		cout << "Ban nhap sai, b phai khac 0" << endl;
		return 0;
	}

	int mask = 1, quotient = 0;

	while ( b <= a){
		b = b << 1;
		mask = mask << 1;
	}

	while ( mask > 1){
		b = b >> 1;
		mask = mask >> 1;

		if( a > b){
			a = a - b;
			quotient = quotient | mask;
		}
	}

	cout << quotient* sign << endl;
}


int main(){
	
	int a, b;
	//cin >> a >> b;
	
	//divide2(a, b)
	divide1(-10000000, 3) ;
	divide2(-10000000, 3);
	divide1(3, 5);
	divide2(3, 5);
	
	return 0;
	
}

File: Week1/Bai1/Bai1.py

# Chia a cho b lấy kết qủa là phần nguyên 
# Xong 
def divide1( a, b):

    if(( a < 0 and b < 0 ) or (a > 0 and b > 0)):
        sign = 1
    else:
        sign = -1
    
    if(b == 0):
        print("Ban nhap sai , b phai khac 0")
        return 

    quotient = 0

    a = abs(a)
    b = abs(b)

    while ( a >= b):
        a = a - b
        quotient = quotient + 1
    
    print(sign*quotient)

def divide2(a, b):

    sign = 1
    if(( a < 0 and b < 0 ) or (a > 0 and b > 0)):
        sign = 1
    else:
        sign = -1
    
    if(b == 0):
        print("Ban nhap sai , b phai khac 0")
        return 


    mask = 1
    quotient = 0

    a =  abs(a)
    b = abs(b)

    while ( b <= a):
        b = b << 1
        mask = mask << 1

    while ( mask > 1):
        b = b >> 1
        mask = mask >> 1

        if ( a > b):
            a = a - b
            quotient = quotient | mask
		

    print(sign * quotient)
	

divide1(10, 3)
divide2(-10, 3)
divide1(0, 3)
divide2(0, 3)
divide1(1, 2)
divide2(3, 0)